cpp/locality/NeighborComputeFunctional.cc

```diff
@@ -15,8 +15,14 @@ NeighborList makeDefaultNlist(const NeighborQuery* nq, const NeighborList* nlist
         auto nqiter(nq->query(query_points, num_query_points, qargs));
         nlist = nqiter->toNeighborList();
     }
+    // Ideally we wouldn't allocate a new NeighborList at all, but we don't want to force
+    // calling code to manage the memory of a raw pointer, so we prefer to return by value
+    // here, and the toNeighborList function has to return a raw pointer to work with the
+    // Cython interface of freud. Therefore, for now we're sticking with this suboptimal
+    // solution until we can refactor these internalsa little more cleanly.
     locality::NeighborList new_nlist = NeighborList(*nlist);
     new_nlist.validate(num_query_points, nq->getNPoints());
+    delete nlist;
```

Comment:
This is only safe if the `nlist` is on the heap, right? Is that guaranteed to be true for all cases where the user is passing in `NeighborList* nlist`? This seems like it would be unsafe if someone did the following, where the `NeighborList*` points to a `NeighborList` on the stack:
```c++
auto nq = ...;
auto my_nlist = NeighborList(...);
freud::locality::makeDefaultNlist(nq, &my_nlist, ...);
```
Examples of this behavior: https://www.geeksforgeeks.org/g-fact-30/

Reply:
You're right that this usage would cause problems with that implementation. In practice that isn't how the function gets used, though. However, this comment made me realize the core problem, which is that we should *never* be deleting a user provided nlist anyway. We only want to delete the nlist we make ourselves here. That should resolve the problems.